cache/warmup.go

```diff
@@ -0,0 +1,66 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/atxs"
+	"github.com/spacemeshos/go-spacemesh/sql/identities"
+	"github.com/spacemeshos/go-spacemesh/sql/layers"
+)
+
+func Warm(db *sql.Database, opts ...Opt) (*Cache, error) {
+	cache := New(opts...)
+	tx, err := db.Tx(context.Background())
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Release()
+	if err := Warmup(tx, cache); err != nil {
+		return nil, fmt.Errorf("warmup %w", err)
+	}
+	return cache, nil
+}
+
+func Warmup(db sql.Executor, cache *Cache) error {
+	latest, err := atxs.LatestEpoch(db)
+	if err != nil {
+		return err
+	}
+	applied, err := layers.GetLastApplied(db)
+	if err != nil {
+		return err
+	}
+	cache.OnApplied(applied.GetEpoch())
+
+	var ierr error
+	if err := atxs.IterateAtxs(db, cache.Evicted(), latest, func(vatx *types.VerifiedActivationTx) bool {
+		nonce, err := atxs.VRFNonce(db, vatx.SmesherID, vatx.TargetEpoch())
+		if err != nil {
+			ierr = fmt.Errorf("missing nonce %w", err)
+			return false
+		}
+		malicious, err := identities.IsMalicious(db, vatx.SmesherID)
+		if err != nil {
+			ierr = err
+			return false
+		}
+		cache.Add(vatx.TargetEpoch(), vatx.SmesherID, vatx.ID(), ToATXData(vatx.ToHeader(), nonce, malicious))
+		return true
+	}); err != nil {
+		return err
+	}
+	return ierr
+}
+
+func ToATXData(atx *types.ActivationTxHeader, nonce types.VRFPostIndex, malicious bool) *ATXData {
```

Comment:
i don't understand this interface used outside the module.
the atx handler needs to query database to see if a node is malicious, and then add the data to cache.
but at the same time the cache also holds whether an node is malicious.

i am confused about the design. a component with access to both, where does it ask for an identity's maliciousness and nonce?

Reply:
consensus doesn't care about malicious identities that doesn't have atx. therefore we load data when atx is received or on startup for previously received atx. you can see how it is used in EligibilityValidator

hare will query cache using GetByNode, and check malicious bit. it is 100% correct, because we really don't care about identity by itself, but only if it has an atx

Reply:
right. the overall plan wasn't really communicated in the PR description.

GetByNode is not used at the moment so i couldn't see its intended purpose.

this is the usage i got form EligibilityValidator
`atx = cache.ToATXData(owned.ToHeader(), nonce, false)`
and i don't know why the malicious bit is always false.